src/DynamoCoreWpf/ViewModels/Core/WorkspaceViewModel.cs

```diff
@@ -1119,6 +1138,61 @@ private void GenerateCombinedGraph()
                     nd.LinkedNotes.Add(note);
                 }
             }
+
+            // Support undo
+            List<ModelBase> undoItems = new List<ModelBase>();
+            undoItems.AddRange(Model.Nodes);
+            undoItems.AddRange(Model.Notes);
+            if (DynamoSelection.Instance.Selection.Count > 0)
+            {
+                undoItems = undoItems.Where(x => x.IsSelected).ToList();
+            }
+            WorkspaceModel.RecordModelsForModification(undoItems, Model.UndoRecorder);
+        }
+
+        private void GenerateCombinedGroupGraph()
+        {
+            Model.LayoutSubgraphs = new List<GraphLayout.Graph>();
+            Model.LayoutSubgraphs.Add(new GraphLayout.Graph());
+
+            GraphLayout.Graph combinedGraph = Model.LayoutSubgraphs.First();
+
+            foreach (NodeModel node in Model.Nodes)
+            {
+                combinedGraph.AddNode(node.GUID, node.Width, node.Height, node.X, node.Y,
+                    node.IsSelected || DynamoSelection.Instance.Selection.Count == 0);
+            }
+
+            foreach (ConnectorModel edge in Model.Connectors)
+            {
+                combinedGraph.AddEdge(edge.Start.Owner.GUID, edge.End.Owner.GUID,
+                    edge.Start.Center.X, edge.Start.Center.Y, edge.End.Center.X, edge.End.Center.Y);
+            }
+
+            foreach (NoteModel note in Model.Notes)
+            {
+                // Link a note to the nearest node
+                GraphLayout.Node nd = combinedGraph.Nodes.OrderBy(node =>
+                    Math.Pow(node.X + node.Width / 2 - note.X - note.Width / 2, 2) +
+                    Math.Pow(node.Y + node.Height / 2 - note.Y - note.Height / 2, 2)).FirstOrDefault();
+
+                if (nd != null)
+                {
+                    nd.LinkedNotes.Add(note);
+                }
+            }
+
+            // Support undo
+            List<ModelBase> undoItems = new List<ModelBase>();
+            foreach (var group in Model.Annotations)
+            {
+                if (group.IsSelected)
+                {
+                    group.Deselect();
+                    undoItems.AddRange(group.SelectedModels);
+                }
+            }
+            WorkspaceModel.RecordModelsForModification(undoItems, Model.UndoRecorder);
         }
```

Comment:
If there is some code here that is common with `GenerateCombinedGraph`, can we factor it out into another private method instead of duplicating the code?